src/world.rs

```diff
@@ -0,0 +1,186 @@
+        let path_str = path.as_ref().to_str().expect("obtaining valid UTF-8 path");
+
+        if let Some(patterns) = &self.patterns {
+            for pattern in patterns {
+                match pattern.match_path(path_str) {
+                    Ok(world_map) => return Ok(world_map),
+                    // We ignore matches here as the path may be matched by another pattern.
+                    Err(Error::NoMatchFound { .. }) => continue,
+                    Err(err) => return Err(err),
+                }
+            }
+        }
+
+        Err(Error::NoMatchFound {
+            path: path.as_ref().to_owned(),
+        })
+    }
+
+    /// Utility function to test a vec of filenames against all defined patterns.
+    /// Returns a vec of results with the parsed [`WorldMap`]s if it matches the pattern.
+    pub fn match_paths<P: AsRef<Path>>(&self, paths: &[P]) -> Vec<Result<WorldMap, Error>> {
+        paths
+            .into_iter()
+            .map(|path| self.match_path(path))
+            .collect()
+    }
+}
+
+/// A WorldMap provides the information for a map in the world and its layout.
+#[derive(Deserialize, PartialEq, Clone, Debug)]
+pub struct WorldMap {
+    /// The filename of the tmx map.
+    #[serde(rename = "fileName")]
+    pub filename: String,
+    /// The x position of the map.
+    pub x: i32,
+    /// The y position of the map.
+    pub y: i32,
+    /// The optional width of the map.
+    pub width: Option<i32>,
+    /// The optional height of the map.
+    pub height: Option<i32>,
+}
+
+/// A WorldPattern defines a regex pattern to automatically determine which maps to load and how to lay them out.
+#[derive(Deserialize, Clone, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct WorldPattern {
+    /// The regex pattern to match against filenames.
+    /// The first two capture groups should be the x integer and y integer positions.
+    #[serde(with = "serde_regex")]
+    pub regexp: Regex,
+    /// The multiplier for the x position.
+    pub multiplier_x: i32,
+    /// The multiplier for the y position.
+    pub multiplier_y: i32,
+    /// The offset for the x position.
+    pub offset_x: i32,
+    /// The offset for the y position.
+    pub offset_y: i32,
+}
+
+impl PartialEq for WorldPattern {
+    fn eq(&self, other: &Self) -> bool {
+        self.multiplier_x == other.multiplier_x
+            && self.multiplier_y == other.multiplier_y
+            && self.offset_x == other.offset_x
+            && self.offset_y == other.offset_y
+            && self.regexp.to_string() == other.regexp.to_string()
+    }
+}
+
+impl WorldPattern {
+    /// Utility function to test a path against this pattern.
+    /// Returns a parsed [`WorldMap`] on the first matched pattern or an error if no patterns match.
+    pub fn match_path(&self, path: impl AsRef<Path>) -> Result<WorldMap, Error> {
+        let path_str = path.as_ref().to_str().expect("obtaining valid UTF-8 path");
```

Comment:
I somewhat doubt the compiler will be able to optimize this check away after it already happened in `World::match_path`. However, since both functions are public API I think they should take a consistent type. Maybe @aleokdev can help decide between `&str` and `AsRef<Path>`?

Reply:
What I've seen in other crates and read in the amazing Rust for Rustaceans book by Jon Gjenset is that, with functions that take a generic parameter which is later monomorphized, two functions are created, one which is public API and exposes the generic parameter (in our case `impl AsRef<Path>` and another which is private and defined either inside the public function's body or alongside it which just takes the type it is converted to (in our case `&str`). That way the function body only gets defined once on the binary, and the only thing that is copied per type input in the generic function is the conversion to &str.

So I think it could be great to apply here. Have the public API function which just calls a `pub(crate) fn match_path_impl` which takes in &str, and use that from World too.